Add compatibility hack for digtron

S-S-X opened the pull request:
Closes #100 
See #100 for more information and possible other options how this could be handled.

I've checked few other mods that add technic machines and did not find similar problems so this might be fine here. Digtron wont actually need that metadata value for anything, for some reason it just checks if it is there.

Not sure about reason behind this one but probably it is to verify that there really is network connected but that should not be needed because machine deactivation ABM should handle this already like it does for other machines and generators.

### Why wont fix:
There's been nasty bug with HV power connector long before any network updates.

> digtron keeps digging for a while after having left cable.

There's no simple fix for that without patching problem in Digtron code, merging this PR makes Digtron power connector work but allows using Digtron without actual power network. If ABM that should disable detached machines wont get executed Digtron will continue as long as it can, that is until it hits lava pool, loses traction, hits ignore nodes or machine timeout ABM executes.

### How this can possibly be fixed without touching Digtron code:
It might be possible to fix this without changing Digtron code if Digtron calls on_construct or on_place when components is moved along with digtron.
If it does so then it might be possible to clear metadata if there's no network connection anymore but this might not work as Digtron does somehow restore previous meta after moving components, also this would be another hacky external patch while Digtron should be able to handle this.
Fixing with callbacks also depend on how Digtron actually restores metadata after moving components.

S-S-X commented:
Should not merge yet, just realized that I possibly used network-ng functions that are not yet available in master branch.
Should check that and maybe also test first...

S-S-X commented:
Updated to not use those function and tested.

SwissalpS commented:
have run some tests. Digtron continues even without cable for a while.
has stopped 'needs more fuel' even when near switching station.

so not quite done yet :)

SwissalpS commented:
Did another two tests on a minimal mod configuration.
Same result: digtron keeps digging for a while after having left cable.

S-S-X commented:
> Digtron continues even without cable for a while.

Have to check how this worked before changing network stuff, I have feeling it has been always somewhat broken especially in case where ABM's get skipped. In laggy conditions engine seems to skip both ABM's and globalsteps. ABM has been way to remove detached machines from network, moving machine should probably actually check for connection after or before every move cycle.

Digtron cycles and network cycles are asynchronous to each other, digtron might execute 2 cycles while network did only 1 cycle. Timeout ABM is worse as it has 1.9 interval and on top of that only 33% chance to get executed:
https://github.com/mt-mods/technic/blob/df93f5d33548a2609cd47ff2063865055a2c2327/technic/machines/switching_station.lua#L604-L609

HV connector keeps full input power until above ABM disables it, that might or might not happen fast enough. Looking at interval and chance most often it does not happen often enough if digtron running 1sec intervals.
This probably should be fixed in digtron mod.

> has stopped 'needs more fuel' even when near switching station.

This seems to be bug in technic network or possibly laggy switching station that gets turned off, probably network problem.

S-S-X commented:
> Digtron continues even without cable for a while.

>  I have feeling it has been always somewhat broken

So tested and this exact same bug happens before network updates, commit 6b3e19e
Not going to even try fixing that here, that is separate issue and it has been probably always broken because of what said in earlier comment. Power connector just have not been used that much because it is not actually useful for mining if working correctly.

> has stopped 'needs more fuel' even when near switching station.

I could not reproduce this one at all and no idea what could cause it, however probably has something to do with adding/removing network nodes. I'll try to make this better with #96

S-S-X commented:
Not sure if digtron HV connector should actually stay disabled until these old problems are fixed, currently if somehow digtron wont crash into wall and lose diggers it could in theory run through whole map if sent away with hv connector, high enough max cycles and timeout ABM does not get executed fast enough.

SwissalpS commented:
ran another test with 5 seconds interval: tron kept going just as far as with 1 second delay.

S-S-X commented:
> ran another test with 5 seconds interval: tron kept going just as far as with 1 second delay.
> Edit: It bugged me enough to run a 15 seconds delay test: same result.

That sound like deactivation through ABM is not happening at all or that we have multiple bugs stacked, both old ones and new ones. I might later test similar around 6b3e19e or even earlier and see what happens. 15 seconds should be enough even with only 33% chance every 2 seconds.

S-S-X commented:
Tested a bit too both 6b3e19e and HEAD of network-ng branch. Behavior is same and as long as area is active and ABM's wont get dropped completely stopping works when Digtron cycle interval is high enough: With 6 seconds interval stopping was bit random but most of time stopped 1-3 nodes from end of cable. 10 second interval always stopped after cable connection was lost.

But ABM is not way to go with Digtron, this cannot be fixed well in technic mod but instead should be fixed in Digtron mod. I am almost sure that exact same bug happens also in upstream technic.

S-S-X commented:
Also verified that this very same bug happens with upstream technic master branch.

BuckarooBanzay commented:
> Only it is not that bad, tron wont usually get very far, I was pretty sure that I know why and so it was. Upstream uses ABM interval 1 second and chance to execute at 100%.

Yeah, that was me a while ago who changed the values on the ABM in a desperate attempt to mitigate some lag issues:

https://github.com/mt-mods/technic/blob/df93f5d33548a2609cd47ff2063865055a2c2327/technic/machines/switching_station.lua#L604-L609

It didn't help much and probably wasn't even the cause, this could be safely reverted back to `1` on both fields IMO...

S-S-X commented:
> It didn't help much and probably wasn't even the cause, this could be safely reverted back to 1 on both fields IMO...

That might still leave some ways to make longer digtron runs without power but will probably not be easy enough or reliable.
So yes agreed that if there's no other reasons to keep ABM less predictable/with longer interval maybe we could restore 1sec/100% chance timeout ABM and use that HV_network meta hack to get around missing metadata.
I'll change that ABM and revert workaround removal.

S-S-X commented:
Added "Fixed" label to indicate that this issue has been partially addressed. However Digtron bugs with HV power connector are not fixed but it should be almost safe with high frequency timeout ABM.
Also reverted removal of digtron `HV_network` meta compatibility workaround.

S-S-X commented:
Other digtron issues do not belong here, this already fixed older bugs. Rest of problems should be handled with separate tickets.
